### ISSUE TITLE
Improving the dialog to get the authentication token

### ISSUE THREAD

**Issue by marco-brandizi** [mguessan/davmail#283]
I've an O365 email account that is managed through the damn InTunes, so I use a chain like Thunderbird - DavMail - O365 (which might require to tweak the client ID, as explained [here](https://sourceforge.net/p/davmail/bugs/702/)).

When I log in from TB, DavMail shows me a dialogue where it tells me to open the URL in which I can go through the corporate login dialogue, I click on 'Open', the web browser shows the usual authentication challenge, I type user/password, then I have another dialogue where I type the authenticator OTP code, and eventually the browser shows a blank page.

But that's not a problem, because I can open the Javascript console and see the log message mentioning the URN that contains the authentication token (see below). I copy-paste such URN back to the DavMail dialogue and click 'Send'. At this point, everything goes on as usual, DavMail talks to O365 and TB gets the last messages.

My question is: can this be made simpler? For instance, is there a way to tell the browser to go back to the DavMail dialogue with the `urn:...` string? Is it possible to implement the dialogue above inside DavMail (eg, using a Java web browser engine)?

![image](https://user-images.githubusercontent.com/534529/232538940-39b73c61-c41b-4696-8e96-58fcbc799e9c.png)

**Comment by mguessan** [comment]
That's exactly what O365Interactive does: proceed with authentication inside an embedded web browser and capture OIDC code to obtain a token.

Unfortunately Java web browser support is limited (relies on OpenJFX) and not compatible with advanced Microsoft authentication mechanisms like Windows Hello for Business.

The only alternative I can see is to have a dedicated Edge based embedded browser (Webview2) with full native authentication support.

**Comment by marco-brandizi** [comment]
Hi @mguessan,

I'm not sure I get it: I'm already using the `O365Interactive` protocol, it doesn't seem to be using any browser.

One thing I see all the time in Chrome and other browsers is that URLs to authenticate in services like Zoom calls or Slack channels (eg, reported in emails)  first open the default browser to offer the authentication dialogue, and then, after successful login, the browser knows how to open the target desktop application. 

There must be some common protocol to manage that, since I see the same behaviour and graphics for different applications and different browsers. Hence, maybe possible solution for DavMail is intercepting those `urn:*` answers thst O365 gives to the browser and re-route them back to DavMail.

**Comment by dngray** [comment]
> Company is now enforcing InTune

How do you know if they are doing that?

> `O365Manual` works but requires same steps with Firebug to copy destination URL

Can you give more details about the steps with Firebug? can this be done with the browser web dev tools? (Ctrl+Shift+I)?

I wanted to use it with the headless docker container [kran0/davmail-docker](https://github.com/kran0/davmail-docker). I made sure to build the container myself against the latest version. The docker container uses `davmail.servers=server` and had to be used with `davmail.mode=O365Manual`. The reason for this is because those libraries installed with azul aren't there. They weren't in the copr repo `mguessan/davmail` either.

In manual mode you're presented with this on the terminal:

> Please open the following link:
> `https://login.microsoftonline.com/common/oauth2/authorize?client_id=d3590ed6-52b3-4102-aeff-aad2292ab01c&response_type=code&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&response_mode=query&login_hint=user%40example.com&resource=https%3A%2F%2Foutlook.office365.com`
> proceed through authentication steps and paste back the final url that contains authentication code (blank page)

What I found is the login page would get stuck and perpetually leave me waiting with the spinny dots at the top.

![2557916268](https://github.com/mguessan/davmail/assets/48640805/12c387a0-acd0-4f61-993b-8b3903b34d8f)

It appears though authentication did happen because if I visit `outlook.office365.com` it logs me straight into the webmail. I don't get redirected to this "blank page" however.

> run `davmail azul` once. It'll download some more libraries and a bit different browser window will be used. Now I can authenticate without any copy/paste steps.

When the mode was set to `davmail.mode=O365Interactive` and `davmail.servers=workstation` the interactive window did appear. When logging into that everything worked and the token was extracted.

I'd like to know what I am missing from the `O365Manual` configuration? Why doesn't it direct me through to this "blank page"?

I did make some wire debug logs but it didn't really reveal what the issue was. I did need [those two options](https://github.com/mguessan/davmail/issues/321#issuecomment-1867072418) as it seems that they won't allow either Thunderbird or Davmail's `client_id`. The [owl extension](https://addons.thunderbird.net/thunderbird/addon/owl-for-exchange/) did work, but I [don't plan on using Thunderbird](https://man.sr.ht/~rjarry/aerc/providers/microsoft.md#office365-with-imap-disabled) beyond testing.

**Comment by dngray** [comment]
> Java web browser support is limited (relies on OpenJFX)

@mguessan By the way I got a warning about that.

> WARNING: A command line option has enabled the GTK 2 library
> WARNING: The JavaFX GTK 2 library is deprecated and will be removed in a future release

Does this mean that the Interactive mode will simply not work on Linux anymore? This was usin gthe embedded JRE "openjdk version "15.0.10" 2023-01-17.

**Comment by cz172638** [comment]
what about borrow something from [linux-entra-sso](https://github.com/siemens/linux-entra-sso)
i tested it with chrome/firefox and i would like to get working davmail.

**Comment by marco-brandizi** [comment]
I am on a macOS. From the little I understand, a simple option might be telling the browser to go to DavMail when a URL has the prefix urn:ietf:wg:auth:oob:*, which is where the MS authentication page redirects the browser after successful authentication. Other apps work this way (eg, Zoom, Slack).

**Comment by mguessan** [comment]
@cz172638  the identity broker seems like an interesting approach, it seems possible to interact with microsoft provided local single sign on broker. This may help on windows workstations that are entra id joined and possibly also on Linux with Intune onboarded workstations.

Even found Microsoft java implementation for this:
https://learn.microsoft.com/en-us/java/api/overview/azure/identity-broker-readme?view=azure-java-stable

Didn't manage to make this work though.

@marco-brandizi it's not that simple: with a custom clientId we can have a http://localhost value as allowed redirecturi, open default browser to let user authenticate and have a listener waiting for the authentication code.

This is how microsoft implemented it in InteractiveBrowserCredential.

The urn: url is typically for embedded authentication (e.g. Teams, Outlook, ...) where the application can intercept returned code.

**Comment by mguessan** [comment]
@erenoglu you may want to try with latest release 6.8.0, ensure you have SWT available either from an OS package or use the following steps:

- Download platform independent package

```
# fetch Azul JDK
davmail azul
# fetch SWT lib
davmail swt
# try to obtain an authentication token
davmail -token
```

SWT does not use the OpenJFX browser but relies on your workstation browser, so if this browser matches Entra conditional access requirements this should work

Same for @marco-brandizi : with SWT no need to use the O365Manual mode

**Comment by marco-brandizi** [comment]
Thanks, @mguessan . This might be useful:

* I had to download the SWT manually, since the URL that the `davmail` script builds doesn't work for macOS
* Even with SWT installed, the auth dialog doesn't pop-up when launching davmail with Temurin JDK 25
* I don't wish to add Azul to my system, but I've tried it anyway. The authentication dialog works inside a Java window, but eventually I get something like: *You can continue only with Devices or client applications that meet `<my-fking-company>` compliance policy*. 
  * That's because the target organisation checks the client ID, and the only way to overcome this is by faking the Outlook ID (they won't accept adding Davmail as a supported client, as they haven't accepted Thunderbird).
  * I still have the Outlook client ID in my settings, but maybe the davmail auth manager doesn't use it.

**Comment by esabol** [comment]
> I had to download the SWT manually, since the URL that the davmail script builds doesn't work for macOS

What should the URL be, @marco-brandizi ?

**Comment by marco-brandizi** [comment]
> > I had to download the SWT manually, since the URL that the davmail script builds doesn't work for macOS
> 
> What should the URL be, @marco-brandizi ?

sorry, already forgotten, search for Eclipse SWT and you'll find it.

**Comment by erenoglu** [comment]
how can I make sure system Edge browser is being used on authentication window? I have the libswt installed on Ubuntu 26.04, but the browser window did not originate from the system Microsoft Edge (which is connected to Intune). So I had to do manual method and after some time token expired.  Other browsers are normally not allowed as they are not compliant with conditional access policy enforced by Intune. whatever I did (even with davmail azul and davmail swt commands), I can't get the system browser. Could there be an option to choose the browser binary for auth?

**Comment by mguessan** [comment]
That's untested, on windows SWT uses webview2 not Edge directly.

To answer your question you can see if SWT is loaded in davmail logs, should appear quickly if you try `davmail -token`

**Comment by erenoglu** [comment]
Thank you. Then this won't work for an Office365 account protected by Intune Conditional Access policies. Ideal solution would be what Evolution or Abraunegg's onedrive client (authenticate once with external browser, then use Intune agent dbus process)

**Comment by erenoglu** [comment]
yes absolutely, I think this is part of the himmelblau project and linux-entra-sso which I use for firefox & chrome.

**Comment by zelch** [comment]
For using himmelblau (or the other native microsoft identity broker), it looks like using https://github.com/siemens/sso-mib might be the easiest path forward.

At least the Linux flatpak is shipping with OpenJDK 21, which gives us access to the Foreign Function & Memory API (FFM) if we enable the preview (google suggests `--enable-preview` for javac, and `--enable-preview --enable-native-access=ALL-UNNAMED` when running the JRE).

And the library API is fairly straight forward from there.

Java isn't a language I've done much in, so I might not be the best choice for making a PR, but I'd be more than happy to test any implementations.

**Comment by mguessan** [comment]
You may also want to investigate msal4j and the python version at:

https://learn.microsoft.com/en-us/entra/msal/python/advanced/linux-broker-py?tabs=ubuntudep

With intune installed the python client should be able to retrieve an access token.
